processing/src/main/java/org/apache/druid/query/operator/GlueingPartitioningOperator.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.operator;
+
+import org.apache.druid.error.InvalidInput;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.query.rowsandcols.ConcatRowsAndColumns;
+import org.apache.druid.query.rowsandcols.LimitedRowsAndColumns;
+import org.apache.druid.query.rowsandcols.RowsAndColumns;
+import org.apache.druid.query.rowsandcols.column.Column;
+import org.apache.druid.query.rowsandcols.column.ColumnAccessor;
+import org.apache.druid.query.rowsandcols.semantic.ClusteredGroupPartitioner;
+import org.apache.druid.query.rowsandcols.semantic.DefaultClusteredGroupPartitioner;
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This glueing partitioning operator is supposed to continuously receive data, and output batches of partitioned RACs.
+ * It maintains a last-partitioning-boundary of the last-pushed-RAC, and attempts to glue it with the next RAC it receives,
+ * ensuring that partitions are handled correctly, even across multiple RACs.
+ * <p>
+ * Additionally, this assumes that data has been pre-sorted according to the partitioning columns.
+ */
+public class GlueingPartitioningOperator extends AbstractPartitioningOperator
+{
+  private final int maxRowsMaterialized;
+  private RowsAndColumns previousRac;
+
+  private static final Integer MAX_ROWS_MATERIALIZED_NO_LIMIT = Integer.MAX_VALUE;
+
+  public GlueingPartitioningOperator(
+      Operator child,
+      List<String> partitionColumns
+  )
+  {
+    this(child, partitionColumns, MAX_ROWS_MATERIALIZED_NO_LIMIT);
+  }
+
+  public GlueingPartitioningOperator(
+      Operator child,
+      List<String> partitionColumns,
+      Integer maxRowsMaterialized
+  )
+  {
+    super(partitionColumns, child);
+    this.maxRowsMaterialized = maxRowsMaterialized;
```

Comment:
could this be `null`?

Reply:
It's `int`, so it can't be null.

Reply:
Makes sense, have added `Preconditions.checkNotNull(maxRowsMaterialized, "maxRowsMaterialized cannot be null");`, hope that works.